### PR TITLE
Fixes DISCO-2493: Add tests for the Suggest component

### PR DESCRIPTION
**Pull request by linabutler**
Some refactoring to make testing easier, and then adding some tests on top of that! It's a little easier to review this commit-by-commit, especially after the first commit that moved the Remote Settings types into their own `rs.rs` (😅) module.

Marking this as a "draft" for now, because there are a few more things (deleting records, records with unknown types, constraining the number of records, clearing) I'd like to cover before landing—but how does the overall shape of this look?

/cc @mtbottle, I can't tag you as a reviewer, but please feel free to review if you'd like!

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/suggest/src/store.rs [hunk 1]
```diff
@@ -71,31 +100,22 @@ pub struct SuggestIngestionConstraints {
...
-    pub fn new(
-        path: &str,
-        settings_config: Option<RemoteSettingsConfig>,
-    ) -> SuggestApiResult<Self> {
-        Ok(Self::new_inner(path, settings_config)?)
-    }
+/// The implementation of the store. This is generic over the Remote Settings
+/// client, and is split out from the concrete [`SuggestStore`] for testing
+/// with a mock client.
+pub(crate) struct GenericSuggestStore<S> {
```
Comment (ncloudioj) [inline components/suggest/src/store.rs:106]: Do we want to bound it?
```suggestion
pub(crate) struct GenericSuggestStore<S: SuggestRemoteSettingsClient> {
```

File: components/suggest/src/store.rs [hunk 2]
```diff
@@ -460,13 +409,469 @@ mod tests {
...
+        );
+        store.ingest(SuggestIngestionConstraints::default())?;
+
+        assert_eq!(
+            store.query(SuggestionQuery {
+                keyword: "la".into(),
+                include_sponsored: true,
+                ..Default::default()
+            })?,
+            &[Suggestion {
```
Comment (ncloudioj) [inline components/suggest/src/store.rs:509]: Looks like we have to write lots of those fixture records in the assertions throughout the file, shall we consider using a snapshot test crate (e.g. [insta](https://github.com/mitsuhiko/insta)) to ease that?

**Review by ncloudioj — APPROVED** [review]
It looks good, nice refactors!

File: components/suggest/src/store.rs (hunk 1 above)
Comment (linabutler) [reply components/suggest/src/store.rs:106]: Good question about adding a bound to the type declaration! I think the Rust convention leans away from doing that (rust-lang/rust-clippy#1689), and prefers adding the bound to just the `impl` blocks that need it. I don't feel strongly about it, though, and would be happy to add the bound if you'd like—the issues with bounds on the type mentioned in that Clippy ticket aren't as relevant here.

You're right about `Generic` being redundant; I had a hard time thinking of a good name! Let's do `SuggestStoreInner` 😁

File: components/suggest/src/rs.rs
Comment (ncloudioj) [inline components/suggest/src/rs.rs:1]: Shall we call it `o.rs`? Then we will get the compiled obj `o.o`

Comment (linabutler) [reply components/suggest/src/rs.rs:1]: 😂😂😂 I _definitely_ didn't call it `rs.rs` intentionally... 😉

File: components/suggest/src/store.rs (hunk 1 above)
Comment (linabutler) [reply components/suggest/src/store.rs]: Yep, that Clippy issue is still open; there's no lint for it or anything now! 👀 But it did help me understand why trait bounds aren't as commonly used, and where they're needed (the `Cow` 🐮 example, where we need the bound to express the `Owned` variant).

**Comment by linabutler** [comment]
I iterated a bit more on the tests, and came up with an approach that I'm a lot happier with!

There's a `Snapshot` helper with convenience constructors that take JSON for the fake records and attachments, and deserialize them into the right data structures. The `SnapshotSettingsClient` just wraps the snapshot, and a nice advantage of making `SuggestStoreInner` generic (vs. using a trait object, from my very first iteration) is that it's super easy to swap out the snapshot.

I'd like to check out https://docs.rs/expect-test/latest/expect_test/ next, and add some more coverage for the store.

File: components/suggest/src/store.rs (hunk 2 above)
Comment (linabutler) [reply components/suggest/src/store.rs]: I added `expect-test` to get us started with snapshot testing, and I'm such a fan! Thanks for the pointer!

**Pull request merged**